File: ABC/ABC176/D.py
import sys
from collections import deque

sys.setrecursionlimit(10 ** 7)
rl = sys.stdin.readline


def solve():
    H, W = map(int, rl().split())
    ch, cw = map(lambda n: int(n) - 1, rl().split())
    dh, dw = map(lambda n: int(n) - 1, rl().split())
    S = [rl().rstrip() for _ in range(H)]
    
    INF = 10 ** 9
    costs = [[INF] * W for _ in range(H)]
    que = deque([(0, ch, cw)])
    while que:
        cost, cy, cx = que.popleft()
        if cy == dh and cx == dw:
            print(cost)
            return
        for dy, dx in ((-1, 0), (0, 1), (1, 0), (0, -1)):
            ny, nx = cy + dy, cx + dx
            if 0 <= ny < H and 0 <= nx < W:
                if S[ny][nx] == '#':
                    continue
                if costs[ny][nx] <= cost:
                    continue
                que.appendleft((cost, ny, nx))
                costs[ny][nx] = cost
        for dy in range(-2, 3):
            for dx in range(-2, 3):
                ny, nx = cy + dy, cx + dx
                if 0 <= ny < H and 0 <= nx < W:
                    if S[ny][nx] == '#':
                        continue
                    if costs[ny][nx] <= cost + 1:
                        continue
                    que.append((cost + 1, ny, nx))
                    costs[ny][nx] = cost + 1
    print(-1)


if __name__ == '__main__':
    solve()

File: ABC/ABC176/D.nim
import deques, math, sequtils, strutils

var
    S: array[1010, string]
    costs: array[1010, array[1010, int]]
    que: Deque[(int, int, int)]


proc solve() =
    var H, W, ch, cw, dh, dw: int
    (H, W) = stdin.readLine.split.map(parseInt)
    (ch, cw) = stdin.readLine.split.map(parseInt)
    (dh, dw) = stdin.readLine.split.map(parseInt)
    dec ch; dec cw; dec dh; dec dw
    for i in 0..<H:
        S[i] = stdin.readLine
    
    for i in 0..<H:
        for j in 0..<W:
            costs[i][j] = 10 ^ 9
    que = initDeque[(int, int, int)]()
    que.addLast((0, ch, cw))

    var cost, cy, cx, ny, nx: int
    while len(que) != 0:
        (cost, cy, cx) = que.popFirst()
        if cy == dh and cx == dw:
            echo cost
            return
        for (dy, dx) in [[-1, 0], [0, 1], [1, 0], [0, -1]]:
            (ny, nx) = (cy + dy, cx + dx)
            if 0 <= ny and ny < H and 0 <= nx and nx < W:
                if S[ny][nx] == '#':
                    continue
                if costs[ny][nx] <= cost:
                    continue
                que.addFirst((cost, ny, nx))
                costs[ny][nx] = cost
        for dy in -2..2:
            for dx in -2..2:
                (ny, nx) = (cy + dy, cx + dx)
                if 0 <= ny and ny < H and 0 <= nx and nx < W:
                    if S[ny][nx] == '#':
                        continue
                    if costs[ny][nx] <= cost + 1:
                        continue
                    que.addLast((cost + 1, ny, nx))
                    costs[ny][nx] = cost + 1
    echo -1


when is_main_module:
    solve()
